core/src/main/java/ai/timefold/solver/core/impl/score/director/EntityValidator.java

```diff
@@ -0,0 +1,183 @@
+            var inverse = inverseRelationShadowVariableDescriptor.getValue(entity);
+            var actualSet = valueToInverse.getOrDefault(entity, Set.of());
+            if (!actualSet.isEmpty() && actualSet.size() != 1) {
+                throw new IllegalStateException(
+                        """
+                                Impossible state: The entity (%s) has a singleton @%s has multiple source entities (%s).
+                                Verify the consistency of your input solution.
+                                """
+                                .formatted(entity, InverseRelationShadowVariable.class.getSimpleName(), actualSet));
+            }
+            if (actualSet.isEmpty() && inverse != null) {
+                throw new IllegalStateException(
+                        """
+                                The entity (%s) has a @%s that points to a source entity (%s) but it has no source entities.
+                                Verify the consistency of your input solution.
+                                """
+                                .formatted(entity, InverseRelationShadowVariable.class.getSimpleName(), inverse));
+            }
+            if (!actualSet.isEmpty()) {
+                var actual = actualSet.iterator().next();
+                if (actual != inverse) {
+                    throw new IllegalStateException(
+                            """
+                                    The entity (%s) has a singleton @%s that points to a source entity (%s) whereas its actual source entity is (%s).
+                                    Verify the consistency of your input solution.
+                                    """
+                                    .formatted(entity, InverseRelationShadowVariable.class.getSimpleName(), inverse,
+                                            actual));
+                }
+            }
+        } else {
+            @SuppressWarnings("unchecked")
+            var inverseCollection = (Collection<Object>) inverseRelationShadowVariableDescriptor.getValue(entity);
+            if (inverseCollection == null) {
+                throw new IllegalStateException("""
+                        The entity (%s) has a collection @%s that is null.
+                        Verify the consistency of your input solution.
+                        """.formatted(entity, InverseRelationShadowVariable.class.getSimpleName()));
+            }
+            var actualCollection = valueToInverse.getOrDefault(entity, Collections.emptySet());
+            if (!inverseCollection.containsAll(actualCollection) || !actualCollection.containsAll(inverseCollection)) {
+                throw new IllegalStateException(
+                        """
+                                The entity (%s) has a collection @%s (%s) that does not match the actual inverse set (%s).
+                                Verify the consistency of your input solution.
+                                """
```

Comment:
This check is very expensive and I wonder if it's worth its weight. Maybe a simpler checking of the sizes would suffice?